### PR TITLE
Eliminate 75GB allocations during binding in a performance trace

### PR DESCRIPTION
**Pull request by sharwell**
Fixes two of the most expensive allocation paths shown in the performance trace attached to [AB#1389013](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1389013).

Fixes [AB#1389013](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1389013)

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs [hunk 1]
```diff
@@ -204,8 +205,12 @@ public virtual ImmutableArray<NamedTypeSymbol> GetTypeMembers(string name, int a
                 {
                     if (syntax != null)
                     {
-                        foreach (var loc in memberT.Locations)
+                        // PERF: Avoid accessing Locations for performance, but assert that the alternative approach is
+                        // equivalent.
+                        Debug.Assert(memberT.MergedDeclaration.Declarations.SelectAsArray(decl => decl.NameLocation).SequenceEqual(memberT.Locations));
+                        foreach (var declaration in memberT.MergedDeclaration.Declarations)
                         {
+                            var loc = declaration.NameLocation;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs:213]: >var loc = declaration.NameLocation; [](http://example.com/codeflow?start=28&length=35)

Is similar optimization applicable to VB?

File: src/Compilers/VisualBasic/Portable/Symbols/LexicalSortKey.vb [hunk 2]
```diff
@@ -207,6 +207,19 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Return LexicalSortKey.Compare(firstKey, secondKey)
         End Function
 
+        Public Shared Function Compare(first As SyntaxNode, second As SyntaxNode, compilation As VisualBasicCompilation) As Integer
```
Comment (jcouv) [inline src/Compilers/VisualBasic/Portable/Symbols/LexicalSortKey.vb:210]: Is this method used?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 2)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

File: src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs (hunk 1 above)
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs:213]: I wasn't able to find a similar method for VB that allows filtering by syntax tree.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs:213]: > I wasn't able to find a similar method for VB that allows filtering by syntax tree.

See methods in ```NamedTypeSymbolExtensions```: ```FindMember```, ```FindFieldOrProperty```, etc.

SemanticModel and BinderFactory often need to find a symbol from its declaration node. It is quite likely location going to be involved in this. The implementation doesn't have to be structured the same way as in C#.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs:213]: I didn't perform exhaustive search

Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/NamespaceOrTypeSymbol.cs:213]: ➡️ Filed https://github.com/dotnet/roslyn/issues/56306

**Pull request merged**